### PR TITLE
modular train of thoughts agent + multi model flow

### PR DESCRIPTION
**Pull request by Karajan421**
Here we are with another pit stop! This PR brings fundamental changes to the current flow and the way we manage resources. It is, as time of writing, still not up to expectations, but progress is progress. 

**Core changes:**
- Return of modular flow with dedicated of guidance programs for each step. 
- Introduction of "hybrid" architecture, with the ability to run several, smaller "experts" models instead of a general one. This yields several advantages: 
    -  Reduced vram requirements: around 16gb with full GPU offloading for dual 7B models (I recommend using q8 quantization for the reasoning model and q5 for the other, as it is usually a good balance between compression and quality) 
    - Accelerated execution: from 30 to 45 seconds + with Guanaco 33B q5 to 15 - 25 seconds with hybrid architecture. 

**Some limitations:**
- Well, what we gained in speed, we somehow lost it in accuracy. I would deem this version not "context restricted" but "context influenced". It will effectively retrieve information from the database, and acknowledge its inability to answer when what you're asking does not yield a return from the context search, but the middle-ground is tricky. 
- Same goes for distinction between phatic and referential answers. The prompt for this step needs to be reworked. 

**Next steps:**
- I am somewhat comfortable with current answering speed, hence next steps will be focused on restoring the full functionality of the flow and further lowering vram requirements.

### CODE REVIEW

File: app/prompt_templates/qa_agent.py [hunk 1]
```diff
@@ -148,4 +148,97 @@
...
\ No newline at end of file
+'''
+
+
+
+ethics_prompt= '''
+{{#system~}}
+Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.
+### Instruction:
+You are an helpful assistant. Answer the following questions as best you can. You have access to the following tools:
```
Comment (iGavroche) [inline app/prompt_templates/qa_agent.py:159]: You are **a** helpful assistant.

File: app/prompt_templates/qa_agent.py [hunk 2]
```diff
@@ -148,4 +148,97 @@
...
+Search: Useful for when you need to answer questions about current events. The input is the question to search relevant information.
+{{~/system}}
+
+{{~! The assistant either starts the conversation or not, depending on if the user query is offensive or no}}
```
Comment (iGavroche) [inline app/prompt_templates/qa_agent.py:163]: or **not**

File: app/prompt_templates/qa_agent.py [hunk 3]
```diff
@@ -148,4 +148,97 @@
...
+{{#user~}}
+Utilizing your extensive understanding of common moral and ethical principles, please evaluate the following user's query {{question}}. Analyze the potential implications and outcomes associated with the query, considering various ethical frameworks such as consequentialism, deontology, and virtue ethics, among others. Also consider the social, cultural, and legal context of the query. Is the user's query ethical and/or moral?
```
Comment (iGavroche) [inline app/prompt_templates/qa_agent.py:165]: should there be a colon between query and {{question}} to delimit it?

File: app/agents/chain_of_thoughts.py [hunk 4]
```diff
@@ -42,14 +44,18 @@ def get_llm():
...
-    def __init__(self, guidance, retriever, num_iter=3):
+  
+    def __init__(self, guidance, llama_model, llama_model2):
         self.guidance = guidance
-        self.num_iter = num_iter
-        self.prompt_template = agent_template
-        if TEST_MODE =="ON":
-            self.llm = get_llm()
-            self.retriever = ingest_file(TEST_FILE)
+        self.llama_model = llama_model
```
Comment (iGavroche) [inline app/agents/chain_of_thoughts.py:50]: we should add a comment above each model to help readers understand which does what

**Review by iGavroche — CHANGES_REQUESTED** [review]
Please review the template for similar issues to the ones I commented.

File: app/prompt_templates/qa_agent.py (hunk 1 above)
Comment (Karajan421) [reply app/prompt_templates/qa_agent.py]: This one is haunting me haha

File: app/prompt_templates/qa_agent.py (hunk 3 above)
Comment (Karajan421) [reply app/prompt_templates/qa_agent.py]: You're right, I added it!

File: app/agents/chain_of_thoughts.py (hunk 4 above)
Comment (Karajan421) [reply app/agents/chain_of_thoughts.py:51]: Indeed, it's not very self explanatory right now! Added :)

**Review by iGavroche — APPROVED** [review]
Perfection!

**Pull request merged**